python/pdo/contract/replication.py

```diff
@@ -0,0 +1,214 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import sys
+import concurrent.futures
+import queue
+import threading
+
+from pdo.contract.state import ContractState
+from pdo.common.utility import are_the_urls_same
+
+import logging
+logger = logging.getLogger(__name__)
+
+class ReplicationException(Exception) :
+    """
+    A class to capture replication exceptions
+    """
+    pass
+
+# -----------------------------------------------------------------
+# -----------------------------------------------------------------
+
+class Replicator(object):
+    """ implements the replicator class functionality : used for change-set replication after contract update, before commiting transaction."""
+    
+    __urls_of_storage_services_to_ignore__ = []
+    __max_num_replicator_threads__ = 2 # max number of replication tasks will be carried out concurrently.
+    __queue_of_SSURLS_raising_exceptions__ = queue.Queue()
+    replication_executor = concurrent.futures.ThreadPoolExecutor(max_workers=__max_num_replicator_threads__) 
+    pending_replications_queue = queue.Queue()
+    exceptions_queue = queue.Queue()
+    replication_completion_condition = threading.Condition()
+    set_of_completed_replications = set()
+   
+    # -----------------------------------------------------------------
+    @classmethod
+    def get_max_num_replicator_threads(cls):
+        return cls.__max_num_replicator_threads__
+
+    # -----------------------------------------------------------------
+
+    @classmethod
+    def replication_worker(cls):
+        """ worker function for a replication task"""
+
+        while True:
+            # get the task from the job queue
+            task = cls.pending_replications_queue.get()
+
+            #check for termination signal
+            if task.get('exit_now'):
+                logger.info("Exiting Replication worker thread")
+                break
+
+            #perform the task
+            try:
+                task['response_object'].replicate_change_set()
+            except Exception as e:
+                cls.exceptions_queue.put(str(e))
+                logger.exception("Replication task failed %s : Exiting Replication worker thread", str(e))
+                break
+
+            #report replication completion in order to add the task to the transaction submission queue
+            task['response_object'].report_replication_completion()
+    
+    # -----------------------------------------------------------------
+
+    @classmethod
+    def __update_set_of_unreliable_storage_services__(cls):
```

Comment:
If i see that correctly, then we ignore permanently any storage service if there was any exception? Isn't that a bit drastic?  Shouldn't we maybe retry after some time, ideally depending on the type of error?

Reply:
the current implementation of the __update_set_of_unreliable_storage_services__ is more of a place holder. One things I wanted to add to this function was to ignore a service if the number of outstanding replication tasks waiting for this service crosses a threshold. And I agree, with you, we could also retry this service "after an exception" after a certain timeout. Once we all agree on a policy, I will make a separate PR for this.

Reply:
High-level answer: "It is super important to isolate the bad ones to avoid thread explosion at the client".

Reason (long one): First of all, not that we do not maintain a single pipe to a storage service to send all packets to this storage service (this is a not a good idea in my opinion, since a storage service might help you with replication attempt now, but might be slow for the next). So if we have 5 storage servicces, for every replication attempt (capped at 2 parallel attempts for now), we start 5 separate threads, one for each storage services. Imagine one ssservice being down, and our request will not timeout for say 5 seconds. if we run 50 updates per seconds, if we do not eliminate the bad sservice, this would mean that we always have a backlog of 250 threads for the entire duration of execution. If we eliminate, then you would have a backlog of 250 threads for just 5 seconds, and after that client only has to deal with threads to the good storage services, thereby gaining significant performance.

Reply:
so maybe rethinking the use of threads would address this problem. it seems like marking bad storage services is the least of the concerns.